### PR TITLE
Create - native desktop windows

### PR DESCRIPTION
**Pull request by thoys**
## Native desktop windows
This PR separates the Create app interface for Desktop in two system-level windows. The Entity List and the Create Tools. The `OverlaySystemWindow` class has been added to the JS API to allow the creation of these system-level windows. 

### QA test plan
1. Start interface in desktop mode
2. Open the create app  ( two windows should appear.  Entity List - left side of the window,  Create Tools - right side of the window)
3. Move the windows around, and resize them. (remember their positions)
4. Switch to VR mode. (The create app should de-activate.)
5. Open the create app. (The create app should function as it does in master, all the create tabs available in tablet   (create, entity list, properties, grid, particle) )
6. Switch back to desktop mode.
7. Open the create app. ( the two windows should appear in their last known locations)
8. Close both desktop windows (Entity List and Create Tools) with the [x].
9. De-activate the create app.
10. Activate the create app. ( the two windows should appear in their last known locations)
11. Smoke test create functionality. (e.g. Create Models, Particles and Shapes)

[edit, jcII: test on win/desk, mac/desk, win/rift, and win/vive. Perform minimum 30 minutes exploratory on each platform after test plan. I'm estimating 90 minutes per platform, 6 hours total. Please tell me if that's too low.]

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10293/HighFidelity-Beta-PR13282-1b107b7.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10293/HighFidelity-Beta-Interface-PR13282-1b107b7.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10309/HighFidelity-Beta-PR13282-b62a5db.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10309/HighFidelity-Beta-Interface-PR13282-b62a5db.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10326/HighFidelity-Beta-PR13282-ffbb2a1.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10326/HighFidelity-Beta-PR13282-ffbb2a1.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10326/HighFidelity-Beta-Interface-PR13282-ffbb2a1.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10326/HighFidelity-Beta-PR13282-ffbb2a1.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10326/HighFidelity-Beta-Interface-PR13282-ffbb2a1.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10332/HighFidelity-Beta-PR13282-7b15a5c.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10332/HighFidelity-Beta-PR13282-7b15a5c.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10332/HighFidelity-Beta-Interface-PR13282-7b15a5c.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10332/HighFidelity-Beta-PR13282-7b15a5c.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10332/HighFidelity-Beta-Interface-PR13282-7b15a5c.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10380/HighFidelity-Beta-PR13282-e96fdec.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10380/HighFidelity-Beta-PR13282-e96fdec.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10380/HighFidelity-Beta-Interface-PR13282-e96fdec.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10380/HighFidelity-Beta-PR13282-e96fdec.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10380/HighFidelity-Beta-Interface-PR13282-e96fdec.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10403/HighFidelity-Beta-PR13282-162d6ef.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10403/HighFidelity-Beta-PR13282-162d6ef.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10403/HighFidelity-Beta-Interface-PR13282-162d6ef.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10403/HighFidelity-Beta-PR13282-162d6ef.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10403/HighFidelity-Beta-Interface-PR13282-162d6ef.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10412/HighFidelity-Beta-PR13282-989791e.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10412/HighFidelity-Beta-PR13282-989791e.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10412/HighFidelity-Beta-Interface-PR13282-989791e.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10415/HighFidelity-Beta-PR13282-2ac1755.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10415/HighFidelity-Beta-PR13282-2ac1755.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10415/HighFidelity-Beta-Interface-PR13282-2ac1755.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10419/HighFidelity-Beta-PR13282-30d5bba.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10420/HighFidelity-Beta-PR13282-b020f63.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10420/HighFidelity-Beta-PR13282-b020f63.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10420/HighFidelity-Beta-Interface-PR13282-b020f63.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10420/HighFidelity-Beta-PR13282-b020f63.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10420/HighFidelity-Beta-Interface-PR13282-b020f63.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10434/HighFidelity-Beta-PR13282-b5cf245.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10434/HighFidelity-Beta-PR13282-b5cf245.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10434/HighFidelity-Beta-Interface-PR13282-b5cf245.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10434/HighFidelity-Beta-PR13282-b5cf245.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10434/HighFidelity-Beta-Interface-PR13282-b5cf245.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10439/HighFidelity-Beta-PR13282-330519a.apk

File: interface/resources/qml/hifi/tablet/EditToolsTabView.qml [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+                        anchors.top: parent.top
+                        anchors.topMargin: 70
+
+
+                        NewEntityButton {
+                            icon: "icons/create-icons/94-model-01.svg"
+                            text: "MODEL"
+                            onClicked: {
+                                editRoot.sendToScript({
+                                                          method: "newEntityButtonClicked", params: { buttonName: "newModelButton" }
```
Comment (dback2) [inline interface/resources/qml/hifi/tablet/EditToolsTabView.qml]: Do these lines need to formatted like this?

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10439/HighFidelity-Beta-PR13282-330519a.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10439/HighFidelity-Beta-Interface-PR13282-330519a.dmg

File: libraries/ui/src/QmlWindowClass.h [hunk 2]
```diff
@@ -24,7 +24,7 @@ class QScriptContext;
  * @param {OverlayWindow.Properties} [properties=null]
  *
  * @hifi-interface
- * @hifi-client-en
+ * @hifi-client-entity
```
Comment (dback2) [inline libraries/ui/src/QmlWindowClass.h]: hifi-client-en is better

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10443/HighFidelity-Beta-PR13282-77448e0.apk

File: scripts/system/edit.js [hunk 3]
```diff
@@ -551,9 +601,21 @@ var toolBar = (function () {
         });
 
         addButton("newModelButton", function () {
-            // tablet version of new-model dialog
-            var tablet = Tablet.getTablet("com.highfidelity.interface.tablet.system");
-            tablet.pushOntoStack("hifi/tablet/NewModelDialog.qml");
+            if (HMD.active) {
```
Comment (dback2) [inline scripts/system/edit.js]: worth this all being a function shared with below?

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10443/HighFidelity-Beta-PR13282-77448e0.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10443/HighFidelity-Beta-Interface-PR13282-77448e0.dmg

File: libraries/ui/src/QmlSystemWindowClass.cpp [hunk 4]
```diff
@@ -0,0 +1,537 @@
...
+    QUrl url{ properties[SOURCE_PROPERTY].toString() };
+    // If the passed URL doesn't correspond to a known scheme, assume it's a local file path
+    if (url.scheme() != "http" && url.scheme() != "https" && url.scheme() != "file" && url.scheme() != "about" &&
+        url.scheme() != "atp" && url.scheme() != "qrc") {
+        properties[SOURCE_PROPERTY] = QUrl::fromLocalFile(url.toString()).toString();
+    }
+
+    return properties;
+}
+namespace hifi { namespace qml { namespace systemWindow {
```
Comment (dback2) [inline libraries/ui/src/QmlSystemWindowClass.cpp]: extra new lines here would be good

File: libraries/ui/src/QmlSystemWindowClass.h [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    QmlSystemWindowClass() = default;
+    ~QmlSystemWindowClass();
+
+    Q_INVOKABLE virtual void initQml(QVariantMap properties);
+
+public slots:
+    bool isVisible();
+    void setVisible(bool visible);
+
+    glm::vec2 getPosition();
```
Comment (dback2) [inline libraries/ui/src/QmlSystemWindowClass.h]: Could be const

File: libraries/ui/src/QmlSystemWindowClass.h [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    void setPosition(const glm::vec2& position);
+    void setPosition(int x, int y);
+
+    glm::vec2 getSize();
```
Comment (dback2) [inline libraries/ui/src/QmlSystemWindowClass.h]: Could be const

File: interface/resources/qml/hifi/tablet/EditToolsTabView.qml [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+        if (typeof id === 'number') {
+            if (id >= 0 && id <= 3) {
+                editTabView.currentIndex = id;
+            } else {
+                console.warn('Attempt to switch to invalid tab:', id);
+            }
+        } else if (typeof id === 'string'){
+            switch (id.toLowerCase()) {
+                case 'create':
+                    editTabView.currentIndex = 0;
```
[the remark this answers is not in the document]
Comment (thoys) [reply interface/resources/qml/hifi/tablet/EditToolsTabView.qml]: No clue how QML could support enums, will try.

File: libraries/ui/src/QmlWindowClass.h (hunk 2 above)
Comment (thoys) [reply libraries/ui/src/QmlWindowClass.h]: I know right!

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10443/HighFidelity-Beta-PR13282-77448e0.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10443/HighFidelity-Beta-Interface-PR13282-77448e0.exe

**Comment by jherico** [comment]
The intent of this PR isn't immediately obvious to me.  I get that it's desirable to be able to use functionliaty like the edit window, entity list, etc, in windows that aren't 'fake windows' inside our application, but rather top level windows...  and this PR seems to be targeting that...  BUT....

* Does this PR make desktop windows in 2D mode and fake windows in VR mode?  Or does it just change a bunch of existing windows to desktop windows, so that they're unusable in VR?
*  It looks like each new window gets it's own QML engine and context created.  Was there an attempt to re-use the engine for the primary UI that didn't work out for some reason?  
* Can we squash these 23 commits into maybe a few or just one?

**Comment by thoys** [comment]
@jherico this is the first phase of many to make the Create App better usable for Creators. 
1. Correct, this PR only uses the desktop windows in Desktop Mode. And HMD friendly windows in VR-mode.
2. Previous attempts to re-purpose the existing engine failed, but now I have a clearer understanding of how QML contexts are created it might be worth it to re-try that. It works as is now. But performance wise it could be better to re-purpose the offscreen QML context. Could be refined in future updates?
3. Definitely. Isn't there an option to squash on merge?

**Comment by thoys** [comment]
3. Just verified. Squash and merge is an option on merge.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10454/HighFidelity-Beta-PR13282-2789346.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10454/HighFidelity-Beta-PR13282-2789346.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10454/HighFidelity-Beta-Interface-PR13282-2789346.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10454/HighFidelity-Beta-PR13282-2789346.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10454/HighFidelity-Beta-Interface-PR13282-2789346.exe

File: libraries/ui/src/QmlSystemWindowClass.cpp [hunk 8]
```diff
@@ -0,0 +1,540 @@
...
+    QUrl url{ properties[SOURCE_PROPERTY].toString() };
+    // If the passed URL doesn't correspond to a known scheme, assume it's a local file path
+    if (url.scheme() != "http" && url.scheme() != "https" && url.scheme() != "file" && url.scheme() != "about" &&
+        url.scheme() != "atp" && url.scheme() != "qrc") {
+        properties[SOURCE_PROPERTY] = QUrl::fromLocalFile(url.toString()).toString();
+    }
+
+    return properties;
+}
+namespace hifi {
```
Comment (dback2) [inline libraries/ui/src/QmlSystemWindowClass.cpp]: Sorry by new lines here I meant spacing lines between the namespace line and the functions. The individual namespaces don't have to be an separate lines since we seem to do this in other places.

File: scripts/system/modules/createWindow.js [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+//  Copyright 2018 High Fidelity, Inc.
+//
+//  Distributed under the Apache License, Version 2.0.
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+
+var getWindowRect = function(settingsKey, defaultRect) {
+    var windowRect = Settings.getValue(settingsKey, defaultRect);
+    return windowRect;
+}
```
Comment (ctrlaltdavid) [inline scripts/system/modules/createWindow.js]: Missing ";"

File: scripts/system/modules/createWindow.js [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+
+var setWindowRect = function(settingsKey, position, size) {
+    Settings.setValue(settingsKey, {
+        position: position,
+        size: size
+    });
+}
```
Comment (ctrlaltdavid) [inline scripts/system/modules/createWindow.js]: Missing ";"

File: scripts/system/modules/createWindow.js [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+        this.qmlPath = qmlPath;
+        this.title = title;
+        this.settingsKey = settingsKey;
+        this.defaultRect = defaultRect;
+        this.webEventReceived = new CallableEvent();
+        this.fromQml = new CallableEvent();
+        if (createOnStartup) {
+            this.createWindow();
+        }
+    };
```
Comment (ctrlaltdavid) [inline scripts/system/modules/createWindow.js]: Extraneous ";"

File: scripts/system/modules/createWindow.js [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+                visible: true,
+                positionX: windowRect.position.x,
+                positionY: windowRect.position.y
+            });
+
+            var windowRectChanged = function() {
+                if (this.window.visible) {
+                    setWindowRect(this.settingsKey, this.window.position, this.window.size);
+                }
+            }
```
Comment (ctrlaltdavid) [inline scripts/system/modules/createWindow.js]: Missing ";"

File: scripts/system/libraries/entityList.js [hunk 13]
```diff
@@ -14,24 +14,61 @@
...
+            return {
+                size: {
+                    x: CREATE_TOOLS_WIDTH,
+                    y: windowHeight
+                },
+                position: {
+                    x: Window.x,
+                    y: Window.y + TITLE_OFFSET
+                }
+            }
```
Comment (ctrlaltdavid) [inline scripts/system/libraries/entityList.js]: Missing ";"

File: scripts/system/libraries/entityList.js [hunk 14]
```diff
@@ -43,18 +80,16 @@ EntityListTool = function(opts) {
             selectedIDs.push(selectionManager.selections[i]);
         }
 
-        var data = {
+        emitJSONScriptEvent({
             type: 'selectionUpdate',
             selectedIDs: selectedIDs,
```
Comment (ctrlaltdavid) [inline scripts/system/libraries/entityList.js]: Extraneous ","

File: interface/resources/qml/hifi/tablet/EntityList.qml [hunk 15]
```diff
@@ -0,0 +1,5 @@
+WebView {
+    id: entityListToolWebView
+    url: Paths.defaultScripts + "/system/html/entityList.html"
+    enabled: true
+}
```
[the remark this answers is not in the document]
Comment (thoys) [reply interface/resources/qml/hifi/tablet/EntityList.qml]: ✔️

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10507/HighFidelity-Beta-PR13282-617441b.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10507/HighFidelity-Beta-Interface-PR13282-617441b.dmg

**Comment by danteruiz** [comment]
Just curious...What is the of the difference between `QmlSystemWIndow` and the `OffscreenUi`? Why couldn't you use `QmlWindow`? Just trying to understand what this new api is used for.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10507/HighFidelity-Beta-PR13282-617441b.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10507/HighFidelity-Beta-Interface-PR13282-617441b.exe

**Comment by thoys** [comment]
@danteruiz OffscreenUI uses QQuickItems as root while the QmlSystemWindow has a QWindow as root QML item, sadly they don't seem to be fully compatible. There are probably things that I could have done differently to make it work more efficiently and make the code more DRY.

File: libraries/ui/src/QmlSystemWindowClass.cpp [hunk 16]
```diff
@@ -0,0 +1,549 @@
...
+QQmlContext* QmlSystemWindowClass::contextForUrl(const QUrl& qmlSource, QObject* parent, bool forceNewContext) {
+    QQmlContext* targetContext = parent ? QQmlEngine::contextForObject(parent) : _qmlContext.data();
+    if (!targetContext) {
+        targetContext = _qmlContext;
+    }
+
+    return targetContext;
+}
+
+#include "QmlSystemWindowClass.moc"
```
Comment (SamGondelman) [inline libraries/ui/src/QmlSystemWindowClass.cpp]: I've discussed this with Dante but this whole class seems like it's redoing a lot of work that's already handled elsewhere.

File: scripts/system/particle_explorer/particleExplorerTool.js [hunk 17]
```diff
@@ -32,17 +33,20 @@ ParticleExplorerTool = function() {
...
         that.activeParticleProperties = {};
 
-        var messageData = {
+        var messageData = JSON.stringify({
             messageType: "particle_close"
-        };
-        that.webView.emitScriptEvent(JSON.stringify(messageData));
+        });
+        that.webView.emitScriptEvent(messageData);
+        createToolsWindow.emitScriptEvent(messageData);
```
Comment (SamGondelman) [inline scripts/system/particle_explorer/particleExplorerTool.js]: I don't like this pattern of duplicating these everywhere...

File: interface/src/Application.cpp [hunk 18]
```diff
@@ -6509,6 +6515,7 @@ void Application::registerScriptEngineWithApplicationServices(ScriptEnginePointe
     scriptEngine->registerFunction("OverlayWebWindow", QmlWebWindowClass::constructor);
 #endif
     scriptEngine->registerFunction("OverlayWindow", QmlWindowClass::constructor);
+    scriptEngine->registerFunction("OverlaySystemWindow", QmlSystemWindowClass::constructor);
```
Comment (danteruiz) [inline interface/src/Application.cpp]: The problem I see with the QmlSystemWindowClass is
1. C++  doesn't actually own this class. We won't can access or manage this window from C++. If the script garbage collects the window, this window will be destroy. (may causes problems)
2. Also you won't be able to access the same window from different scripts, which might be nice.

File: interface/resources/qml/QmlSystemWindow.qml [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+        Qt.WindowTitleHint |
+        Qt.WindowSystemMenuHint |
+        Qt.WindowCloseButtonHint |
+        Qt.WindowMaximizeButtonHint |
+        Qt.WindowMinimizeButtonHint
+    color: hifi.colors.baseGray
+    resizable: true
+    property var channel;
+    // Don't destroy on close... otherwise the JS/C++ will have a dangling pointer
+    destroyOnCloseButton: false
```
[the remark this answers is not in the document]
Comment (thoys) [reply interface/resources/qml/QmlSystemWindow.qml]: Yes, I handle the close of the window in script using the close() function. Similar as how OverlayWindow / QmlWindow does it. That comment is directly copied from QmlWindow.qml

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10531/HighFidelity-Beta-PR13282-782f5b7.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10531/HighFidelity-Beta-PR13282-782f5b7.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10531/HighFidelity-Beta-Interface-PR13282-782f5b7.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10531/HighFidelity-Beta-PR13282-782f5b7.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10531/HighFidelity-Beta-Interface-PR13282-782f5b7.exe

File: interface/src/Application.cpp [hunk 20]
```diff
@@ -3019,10 +3018,7 @@ void Application::setupQMLContext(QQmlContext* surfaceContext) {
     if (auto steamClient = PluginManager::getInstance()->getSteamClientPlugin()) {
         surfaceContext->setContextProperty("Steam", new SteamScriptingInterface(engine, steamClient.get()));
     }
-}
 
-void Application::onDesktopRootContextCreated(QQmlContext* surfaceContext) {
-    setupQMLContext(surfaceContext);
```
Comment (danteruiz) [inline interface/src/Application.cpp]: Why did you remove this?

File: libraries/qml/src/qml/OffscreenSurface.h [hunk 21]
```diff
@@ -115,7 +117,7 @@ public slots:
     virtual void onItemCreated(QQmlContext* context, QQuickItem* newItem) {}
     virtual void onRootContextCreated(QQmlContext* qmlContext) {}
 
-    virtual QQmlContext* contextForUrl(const QUrl& qmlSource, QQuickItem* parent, bool forceNewContext);
+    virtual QQmlContext* contextForUrl(const QUrl& qmlSource, QQuickItem* parent, bool forceNewContext, bool useParentContext);
 private:
```
Comment (danteruiz) [inline libraries/qml/src/qml/OffscreenSurface.h]: I don't quite understand why you had to change the process we load qml files? Can you explain the changes in this file

File: interface/src/Application.cpp (hunk 20 above)
Comment (thoys) [reply interface/src/Application.cpp]: Because I added it

File: libraries/ui/src/QmlSystemWindowClass.cpp [hunk 22]
```diff
@@ -150,33 +98,38 @@ QScriptValue QmlSystemWindowClass::constructor(QScriptContext* context, QScriptE
 void QmlSystemWindowClass::initQml(QVariantMap properties) {
     _source = properties[SOURCE_PROPERTY].toString();
 
+    auto offscreenUi = DependencyManager::get<OffscreenUi>();
+    _qmlWindow = new QQuickView(offscreenUi->getSurfaceContext()->engine(), nullptr);
```
Comment (danteruiz) [inline libraries/ui/src/QmlSystemWindowClass.cpp]: Why not just use a new instance of the `OffscreenQmlSurface` directly instead of going through the OffscreenUI? What there an issue you encountered?

Comment (danteruiz) [inline libraries/ui/src/QmlSystemWindowClass.cpp]: Why not just use a new instance of the `OffscreenQmlSurface` directly instead of going through the OffscreenUI? What there an issue you encountered?

File: interface/src/Application.cpp (hunk 20 above)
Comment (danteruiz) [reply interface/src/Application.cpp]: oh ok

File: libraries/qml/src/qml/OffscreenSurface.h (hunk 21 above)
Comment (thoys) [reply libraries/qml/src/qml/OffscreenSurface.h]: Yes, if I don't set the parent on creation then the item got stuck to the main window UI. That's why I send the QQuickView->contentItem() as parent. But I do not want to use this "parent" as the context because that would mean that it has a blank context with non of the hifi classes available, so `useParentContext = false` in this case (`true` by default).

File: libraries/ui/src/QmlSystemWindowClass.cpp (hunk 22 above)
Comment (thoys) [reply libraries/ui/src/QmlSystemWindowClass.cpp]: Not sure, I followed the example in QmlWindowClass.cpp on how to create a surface. Is there a problem with this approach? Would I otherwise have to use the OffscreenQmlSurfaceCache? That might cause side-effects?

File: libraries/qml/src/qml/OffscreenSurface.h (hunk 21 above)
Comment (danteruiz) [reply libraries/qml/src/qml/OffscreenSurface.h]: You might be able to avoid all this if you used a separate OffscreenQmlSurface.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10544/HighFidelity-Beta-PR13282-c336845.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10544/HighFidelity-Beta-PR13282-c336845.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10544/HighFidelity-Beta-Interface-PR13282-c336845.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10544/HighFidelity-Beta-PR13282-c336845.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10544/HighFidelity-Beta-Interface-PR13282-c336845.exe

File: interface/src/Application.cpp (hunk 18 above)
Comment (jherico) [reply interface/src/Application.cpp]: it's possible to explicitly tell QScriptEngine that an object is owned by C++ code, so it shouldn't be deleted, and there are places where that's the default, although the exact circumstances elude me at the moment.  

I don't think having windows shared between scripts is a good idea.  If there's a need for script coordination, we should explicitly meet that need, not just do a what-if on the idea of multiple scripts sharing a single window.
